File: zig/day03/solve.zig
const std = @import("std");
const util = @import("../util.zig");

const BITS = 12;
const Bits = u12;
const BitIndex = u4;

fn parseBitstrings(mem: std.mem.Allocator, input: []const u8) ![]Bits {
  var nums = std.ArrayList(Bits).init(mem);
  errdefer nums.deinit();

  var iter = std.mem.tokenize(u8, input, "\n");
  while (iter.next()) |line| {
    const n = try std.fmt.parseUnsigned(Bits, line, 2);
    try nums.append(n);
  }

  return nums.toOwnedSlice();
}

fn bitAt(bits: Bits, index: BitIndex) u1 {
  return @intCast(u1, bits >> (BITS - index - 1) & 1);
}

fn mostCommonBit(nums: []Bits, bit_index: BitIndex) u1 {
  var count: usize = 0;
  for (nums) |n| count += bitAt(n, bit_index);
  return @boolToInt(2 * count >= nums.len);
}

fn gammaEpsilon(nums: []Bits) [2]Bits {
  var gamma: Bits = 0;
  var bit_index: BitIndex = 0;
  while (bit_index < BITS) : (bit_index += 1) {
    gamma = gamma << 1 | mostCommonBit(nums, bit_index);
  }
  return .{ gamma, ~gamma };
}

fn findByCriterion(mem: std.mem.Allocator, nums: []Bits, by_most_common: bool) !Bits {
  var state = std.ArrayList(Bits).init(mem);
  defer state.deinit();

  try state.appendSlice(nums);

  var bit_index: BitIndex = 0;
  while (bit_index < BITS) : (bit_index += 1) {
    const most_common = mostCommonBit(state.items, bit_index);
    const bit = most_common ^ @boolToInt(by_most_common);

    var index = state.items.len;
    while (index > 0) {
      index -= 1;
      if (bitAt(state.items[index], bit_index) != bit)
        _ = state.swapRemove(index);
    }

    if (state.items.len == 1)
      return state.items[0];
  }

  unreachable;
}

pub fn solve(mem: std.mem.Allocator) !void {
  const input = try util.readFileAlloc(mem, ".input/day03");
  defer mem.free(input);

  const nums = try parseBitstrings(mem, input);
  defer mem.free(nums);

  const gamma_epsilon = gammaEpsilon(nums);
  const oxygen = try findByCriterion(mem, nums, true);
  const co2 = try findByCriterion(mem, nums, false);

  const part1 = @as(u32, gamma_epsilon[0]) * @as(u32, gamma_epsilon[1]);
  const part2 = @as(u32, oxygen) * @as(u32, co2);

  std.debug.print("day03: {} {}\n", .{ part1, part2 });
}

File: c++/day03/solve.cpp
#include <array>
#include <vector>
#include "../util.hpp"

namespace day03 {
  const int BITS = 12;
  const int MASK = (1 << BITS) - 1;

  auto parse_bitstrings(std::string_view input) -> std::vector<int> {
    std::vector<int> nums;
    for (auto line : util::tokenize(input)) {
      auto bits = 0;
      for (auto bit : line)
        bits = bits << 1 | bit - '0';
      nums.emplace_back(bits);
    }
    return nums;
  }

  auto bit_at(int bits, int index) -> bool {
    return bits >> (BITS - index - 1) & 1;
  }

  auto most_common_bit(std::vector<int>& nums, int bit_index) -> bool {
    auto count = 0;
    for (auto bits : nums)
      count += bit_at(bits, bit_index);
    return 2 * count >= nums.size();
  }

  auto gamma_epsilon(std::vector<int>& nums)->std::array<int, 2> {
    auto gamma = 0;
    auto bit_index = 0;
    for (auto bit_index = 0; bit_index < BITS; bit_index += 1)
      gamma = gamma << 1 | most_common_bit(nums, bit_index);
    return { gamma, gamma ^ MASK };
  }

  auto find_by_criterion(std::vector<int> nums, bool by_most_common) -> int {
    for (auto bit_index = 0; bit_index < BITS; bit_index += 1) {
      auto most_common = most_common_bit(nums, bit_index);
      auto bit = most_common ^ by_most_common;

      for (auto index = nums.size(); index--;) {
        if (bit_at(nums[index], bit_index) != bit) {
          nums[index] = nums.back();
          nums.pop_back();
        }
      }

      if (nums.size() == 1)
        return nums[0];
    }

    throw std::runtime_error{ "bruh moment" };
  }

  auto solve() -> void {
    auto input = util::read_file_to_string(".input/day03");
    auto nums = parse_bitstrings(input);

    auto [gamma, epsilon] = gamma_epsilon(nums);
    auto oxygen = find_by_criterion(nums, true);
    auto co2 = find_by_criterion(nums, false);

    auto part1 = gamma * epsilon;
    auto part2 = oxygen * co2;

    std::cout << "day03: " << part1 << ' ' << part2 << '\n';
  }
}
